Transform async `ResumeTy` in generator transform

Swatinem opened the pull request:
- Eliminates all the `get_context` calls that async lowering created.
- Replace all `Local` `ResumeTy` types with `&mut Context<'_>`.

The `Local`s that have their types replaced are:
- The `resume` argument itself.
- The argument to `get_context`.
- The yielded value of a `yield`.

The `ResumeTy` hides a `&mut Context<'_>` behind an unsafe raw pointer, and the `get_context` function is being used to convert that back to a `&mut Context<'_>`.

Ideally the async lowering would not use the `ResumeTy`/`get_context` indirection, but rather directly use `&mut Context<'_>`, however that would currently lead to higher-kinded lifetime errors.
See <https://github.com/rust-lang/rust/issues/105501>.

The async lowering step and the type / lifetime inference / checking are still using the `ResumeTy` indirection for the time being, and that indirection is removed here. After this transform, the generator body only knows about `&mut Context<'_>`.

---

Fixes https://github.com/bjorn3/rustc_codegen_cranelift/issues/1330 CC @bjorn3 

r? @compiler-errors

Swatinem committed:
Transform async ResumeTy in generator transform

- Eliminates all the `get_context` calls that async lowering created.
- Replace all `Local` `ResumeTy` types with `&mut Context<'_>`.

The `Local`s that have their types replaced are:
- The `resume` argument itself.
- The argument to `get_context`.
- The yielded value of a `yield`.

The `ResumeTy` hides a `&mut Context<'_>` behind an unsafe raw pointer, and the
`get_context` function is being used to convert that back to a `&mut Context<'_>`.

Ideally the async lowering would not use the `ResumeTy`/`get_context` indirection,
but rather directly use `&mut Context<'_>`, however that would currently
lead to higher-kinded lifetime errors.
See <https://github.com/rust-lang/rust/issues/105501>.

The async lowering step and the type / lifetime inference / checking are
still using the `ResumeTy` indirection for the time being, and that indirection
is removed here. After this transform, the generator body only knows about `&mut Context<'_>`.

compiler-errors commented:
I'm probably not the best person to review mir-opt work.

r? @rust-lang/wg-mir-opt

bors commented:
:umbrella: The latest upstream changes (presumably #105741) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

eddyb commented:
> @rustbot assigned @eddyb and unassigned @compiler-errors (2 weeks ago)

Uhh that's a bug I'm pretty sure? I shouldn't be given any reviews (cc @oli-obk).

r? @rust-lang/wg-mir-opt

oli-obk commented:
> Uhh that's a bug I'm pretty sure? I shouldn't be given any review

explicit team assignments will pick anyone from that team, even if they are not on the automatic rotation

spastorino commented:
r? @rust-lang/wg-mir-opt

oli-obk commented on compiler/rustc_mir_transform/src/generator.rs:
```diff
@@ -1308,8 +1408,21 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
-            replace_local(resume_local, body.local_decls[resume_local].ty, body, tcx);
+        let original_resume_ty = body.local_decls[resume_local].ty;
+        let resume_ty = if is_async_kind {
+            let context_did = tcx.require_lang_item(LangItem::Context, None);
+            let context_adt_ref = tcx.adt_def(context_did);
+            let context_substs = tcx.intern_substs(&[tcx.lifetimes.re_erased.into()]);
+            let context_ty = tcx.mk_adt(context_adt_ref, context_substs);
+            let context_mut_ref = tcx.mk_mut_ref(tcx.lifetimes.re_erased, context_ty);
```
maybe add a `TyCtxt` method for this snippet and reuse it in abi.rs

oli-obk replied:
also, this could be called within `transform_async_context`

oli-obk commented on compiler/rustc_mir_transform/src/generator.rs:
```diff
@@ -460,6 +460,106 @@ fn replace_local<'tcx>(
+/// The async lowering step and the type / lifetime inference / checking are
+/// still using the `ResumeTy` indirection for the time being, and that indirection
+/// is removed here. After this transform, the generator body only knows about `&mut Context<'_>`.
+fn transform_async_context<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &mut Body<'tcx>,
+    context_mut_ref: Ty<'tcx>,
+) {
```
I wonder if we could register this as a special mir opt so we can see its diff in a mir opt test. Seems like a lot of work though, and out of scope for this PR

Swatinem replied:
That would be interesting, but not sure if its worth it.

I consider this PR just a temporary workaround that I would love to revert as soon as the compiler can handle `&mut Context<'_>` in generators directly.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 1db48e2b284530513cf825f6c2dc3f573a0cad81 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 1db48e2b284530513cf825f6c2dc3f573a0cad81 -->
<!-- homu: {"type":"Approved","sha":"1db48e2b284530513cf825f6c2dc3f573a0cad81","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
Needs mir-opt tests blessed: https://github.com/rust-lang/rust/pull/107046#issuecomment-1396204564

@bors r-

Swatinem commented:
Rebased, and added `compile-flags: -C panic=abort` to the test.
The problem was that the `wasm32` target defaults to `panic=abort` and thus no unwinding paths and `future polled after panic` was generated there.
Using `panic=abort` directly should make the output consistent across targets and even simpler.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 96931a787abaac7d720fc415780037fc63bee98d has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 96931a787abaac7d720fc415780037fc63bee98d -->
<!-- homu: {"type":"Approved","sha":"96931a787abaac7d720fc415780037fc63bee98d","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented:
Updated cg_clif in https://github.com/bjorn3/rustc_codegen_cranelift/commit/6c58be8e518e311d9769c595adad4bd30e65aff6. Works perfect!
